Update snyk iac cli to issue parallel requests

ipapast opened the pull request:
- [ ] Ready for review
- [ ] Follows [CONTRIBUTING](https://github.com/snyk/snyk/blob/master/.github/CONTRIBUTING.md) rules
- [ ] Reviewed by Snyk internal team

#### What does this PR do?

Currently for a call to snyk iac test our CLI process makes the following sequential requests to our backend API:
- One request per file to validate each Terraform file in the directory tree (or provided via arguments)
- One request pre file to test each matching IaC file in the directory tree (or provided via arguments)

1. As part of cloud-config's work on making tthe cli more performant, we are changing the requests above to **run in parallel.** Various tests were made to determine how this affects the timings, with results showing an average decrease of 8m30s. This is a small win which makes the cli run much faster.
2. In order not to throttle the system in the case of a big number of files being tested, we also introduced two small changes. By default, HTTP connections close after each request. The keepAlive property will allow a persistent connection. For rate limiting, a queue was implemented for both detecting the iac files and testing them. 

The results look very promising. From the original 10m22sec when run locally, it went down to 1min4s with a queue.

This is a table with some tests. It was run against cloud-config-goof, which contains 249 IaC files.

_Tested 247 projects, 201 contained issues. Failed to test 4 projects._

The goal of these experiments was to see what the "optimal" number of concurrent requests is for our case.
- The "before" run is by running it locally without any changes from the master branch. 
- The "after" run is by running it locally against the proposed changes.

***table to be updated asap***
Scenario | user | system | cpu | Total
-- | -- | -- | -- | --
Before | 14.32s | 36.63s | 8% | 10m28s18ms
After: |  
with promises running in parallel | 12.00s | 31.08s | 52% | 1m22s71ms
with promises running in parallel & Queue | 28.00s | 84.08s | 92% | 1m12s71ms
with  rest && https.agent {keepAlive: true} | 14.49s | 43.21s | 51% | 1m52s74ms

#### How should this be manually tested?
- Run the code on the master branch:
`time node snyk iac test`
- Then go to this branch and run `time node ~/snyk-repos/snyk/dist/cli iac test` .
Note the 'total' time in the end. It should have been improved.

#### Any background context you want to provide?
There were [more performance improvements](https://snyksec.atlassian.net/jira/software/projects/CC/boards/93?selectedIssue=CC-519) done on the cloud-config-policy-engine where a new endpoint and a feature flag were added. 
As part of the current task, we also use those changes in combination. 
To do this, we needed to enable the _iacEvaluatePolicyEndpoint_ flag in the local snyk admin area when running locally. 

#### What are the relevant tickets?
https://snyksec.atlassian.net/browse/CC-594

ipapast commented on src/lib/snyk-test/run-test.ts:
```diff
@@ -292,6 +298,11 @@ async function sendAndParseResults(
     }
   }
+
+  const setTimeoutAsync = util.promisify(setTimeout);
+  while (queue.getQueueLength() + queue.getPendingLength() > 0) {
+    await setTimeoutAsync(100);
+  }
```
The library doesn't provide any built in function yet, so this workaround was used instead. Inspired by an [open issue](https://github.com/promise-queue/promise-queue/issues/23)

ipapast commented on src/lib/iac/detect-iac.ts:
```diff
@@ -107,31 +111,31 @@ async function getDirectoryFiles(
   }
 
-  const iacFiles = await Promise.all(promises);
+  const setTimeoutAsync = util.promisify(setTimeout);
+  while (queue.getQueueLength() + queue.getPendingLength() > 0) {
+    await setTimeoutAsync(100);
+  }
```
The library doesn't provide any built in function yet, so this workaround was used instead.

almog27 replied:
If so - not worth considering something that do have the async ability? such as `async-promise-queue`?

aron replied:
In this instance we know all the promises that we are trying to make requests for. The `add()` method returns a promise that resolves to the result of the enqueued function so we can make use of that.

```
// In the loop...
promises.push(queue.add(() => ...));

// Here:
if (promises.length > 0) {
  return await Promise.all(promises);
}
```

In case that doesn't make sense :) I've put a demo together here: https://runkit.com/aron/promisequeue-example

almog27 commented on src/lib/iac/detect-iac.ts:
```diff
@@ -90,14 +92,16 @@ async function getDirectoryFiles(
   });
 
+  const iacFiles: IacFileInDirectory[] = [];
+  const queue = new Queue(50);
```
Worth explaining what is 50 by adding some const - my first thought was a queue max size and not the max concurrent

aron commented on src/lib/snyk-test/run-test.ts:
```diff
@@ -226,23 +226,36 @@ async function sendAndParseResults(
+            projectName,
+            options.severityThreshold,
+          );
+          results.push(result);
```
The results are now captured in the promises array so we could just return `result` here and drop the additional `results` array.

aron commented on src/lib/snyk-test/run-test.ts:
```diff
@@ -226,23 +226,36 @@ async function sendAndParseResults(
     await spinner.clear<void>(spinnerLbl)();
     await spinner(spinnerLbl);
     if (options.iac) {
-      const iacScan: IacScan = payload.body as IacScan;
```
A couple of extra observations here now.

1) Previously this would reset the spinner once per iteration. As we're now batching things this is not going to work as expected...
1) I think we could simplify this by handling the IaC condition first, otherwise handling the other paths rather than doing it inside the loop. This would mean that we could declare `promises` for IaC only and `results` for the primary path. This would aid refactoring in the future allowing us to extract the iac path into a seperate function.

Example:

```ts
// IaC path...
if (options.iac) {
  const promises: Promise<TestResult>[] = []
  const maxConcurrent = 25;
  const queue = new Queue(maxConcurrent);

  await spinner(spinnerLbl);
  for (const payload of payloads) {
    promises.push(queue.add(...));
  }

  return Promise.all(promises)
}

// Standard path...
const results: TestResult[] = []

...

for (const payload of payloads) {
    await spinner.clear<void>(spinnerLbl)();
    await spinner(spinnerLbl);
    ...
}

return results
```

aron reviewed (commented):
This is looking really good from my side, loving those time improvements ❤️ . A couple of final comments around typings and code structure. Also regarding the commit messages/release notes the only one that is really a "feature" is "run validate & test in parallel" can we update the other two to use the `refactor:` label. And let's include "IaC" in the othere one for clarity e.g. "Run the IaC queries in validate & test in parallel"

ipapast commented on src/lib/snyk-test/run-test.ts:
```diff
@@ -226,23 +226,36 @@ async function sendAndParseResults(
   options: Options & TestOptions,
 ): Promise<TestResult[]> {
   const results: TestResult[] = [];
+  const promises: any[] = [];
```
updated to `Array<Promise<TestResult>> = [];`

aron commented on test/request.test.ts:
```diff
@@ -49,7 +51,7 @@ test('request calls needle as expected and returns status code and body', (t) =>
             timeout: 300000, // default
             json: undefined, // default
-            agent: undefined, // should not be set when not using proxy
+            agent: new https.Agent({ keepAlive: true }), // should not be set when not using proxy
```
nit: comment is out of date.

aron commented on src/lib/request/request.ts:
```diff
@@ -94,19 +95,24 @@ export = function makeRequest(
         if (proxyUri) {
           snykDebug('using proxy:', proxyUri);
-          options.agent = (new ProxyAgent(proxyUri) as unknown) as Agent;
+          options.agent = (new ProxyAgent(proxyUri) as unknown) as http.Agent;
```
Lets also install the types fro ProxyAgent via `npm install @types/http-proxy-agent` and avoid this casting.

aron commented on src/lib/iac/detect-iac.ts:
```diff
@@ -90,14 +91,18 @@ async function getDirectoryFiles(
+  const maxConcurrent = 50;
+  const queue = new Queue(maxConcurrent);
+
+  const promises: Array<Promise<IacFileInDirectory>> = [];
```
nit: The `type[]` is a more common ideom now for expressing an array.

```suggestion
  const promises: Promise<IacFileInDirectory>[] = [];
```

ipapast replied:
Ah, this was my auto TSLint fix.
Apparently it is giving an: 
`TSLint: Array type using 'T[]' is forbidden for non-simple types. Use 'Array<T>' instead.(array-type)` [link](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/array-type.md)

From a bit of googling, I understand that **[ ]** and **Array** are the same, except for the times we are using a `readonly` type. [reference](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-4.html#caveats)

I'm happy to put it back as it was since this is the preference, just wondering if I should be adding an ignore rule in TSLint too:
`"rules": {
    "array-type": false
}`

aron replied:
Oh my apologies, leave it as is.

aron commented on src/lib/iac/detect-iac.ts:
```diff
@@ -106,27 +111,33 @@ async function getDirectoryFiles(
+      queue.add(async () => {
+        try {
+          const projectType = await getProjectTypeForIacFile(filePath);
+          iacFiles.push({
```
Same cleanup applies here, we no longer need both `promises` and `iacFiles` we can just return the object here and resolve promises later on to get the files.

ipapast replied:
🤦 thanks

aron commented on src/lib/iac/detect-iac.ts:
```diff
@@ -106,27 +111,33 @@ async function getDirectoryFiles(
-          err instanceof CustomError ? err.userMessage : 'Unhandled Error',
-      });
-    }
+  if (promises.length > 0) {
```
Suggestion:

```ts
if (promises.length === 0) {
  throw IacDirectoryWithoutAnyIacFileError();
}
return Promise.all(promises);
```

aron commented on src/lib/snyk-test/run-test.ts:
```diff
@@ -225,72 +225,90 @@ async function sendAndParseResults(
+  if (options.iac) {
+    const maxConcurrent = 25;
+    const queue = new Queue(maxConcurrent);
+    const iacResults: Array<Promise<TestResult>> = [];
```
nit:

```suggestion
    const iacResults: Promise<TestResult>[] = [];
```

ipapast replied:
^^I'm bringing it back too

aron reviewed (approved):
👏 This looks great, a huge improvement on the previous test times. I've a few minor comments below where I think things would benefit from a little more cleanup but otherwise I think this is ready for review from the wider team.

JackuB commented on src/lib/request/request.ts:
```diff
@@ -5,12 +5,13 @@ import * as querystring from 'querystring';
 import * as config from '../config';
 import { getProxyForUrl } from 'proxy-from-env';
-import * as ProxyAgent from 'proxy-agent';
+import HttpProxyAgent = require('http-proxy-agent');
```
Two things here:

1. `proxy-agent` is not used anymore, so it may be removed from the package.json?
2. `http-proxy-agent` has types defined in package.json, but not the package itself?

And out of curiosity, why the different import/require statement?

ipapast replied:
1. good catch, removed 👍 

About the import/require: 
HttpProxyAgent is exported like this in @types/http-proxy-agent:
`export = HttpProxyAgent;`
And when I was trying a traditional i`mport {HttpProxyAgent} from '...'` I was experiencing errors, so I found in the documentation that the `import HttpProxyAgent = require('http-proxy-agent');` is what we should be using when something is exported as `export =`.  You can see it [here](https://www.typescriptlang.org/docs/handbook/modules.html#export--and-import--require)

aron replied:
Oh this is my fault I think. @ipapast [`proxy-agent`](https://www.npmjs.com/package/proxy-agent) and [`http-proxy-agent`](https://www.npmjs.com/package/http-proxy-agent) are two different libraries that do the same thing maintained by the same person. `proxy-agent` supports more than just HTTP. Because I don't have a full understanding of how this is used downstream I think we need to keep using `proxy-agent` which also seems to bundle types but it may be that they're incompatible with `http.Agent`. So I think the original typing issue may be resolved with just an upgrade, or it might be that we need to keep the awkward casting.

ipapast replied:
So I've been trying to make it work with `http-proxy-agent`. IT WON'T. 😢

Since the **http-proxy-agent** module uses old-school imports like: 
```
import net from 'net';
import tls from 'tls';
```

When we run the build,  it will fail with errors of style 
```
Module '"net"' has no default export.
Module '"tls"' has no default export.
```
Unfortunately, it's a third party module so we can't really edit these imports. The only way to make it pass was by adding `"allowSyntheticDefaultImports": true` in the tsconfig.json

ipapast replied:
Upgrading proxy-agent to 4.0.1 (latest) leads to very similar errors which are coming from a chained import of the `agent-base`.
```
 Module '"net"' has no default export.
 Module '"http"' has no default export.
  Module '"https"' has no default export.
```

ipapast replied:
Reverted back to http-proxy-agent and the weird casting. Added comment: `proxyAgent type is an EventEmitter and not an http Agent`

JackuB commented on src/lib/snyk-test/run-test.ts:
```diff
@@ -225,72 +225,90 @@ async function sendAndParseResults(
-        options,
-      );
+    /** sendTestPayload() deletes the request.body from the payload once completed. */
+    const payloadCopy = Object.assign({}, payload);
```
Is it problematic here that `Object.assign` is not creating a deep clone?

ipapast replied:
hm not sure, but maybe @almog27 can help. do we need a deep clone of the payload or is it ok as it is?

aron replied:
Worth noting that this is existing behavior and only shows up because it's now outside of the `else` block.

JackuB commented on src/lib/iac/detect-iac.ts:
```diff
@@ -90,14 +91,17 @@ async function getDirectoryFiles(
   });
 
+  const iacFiles: IacFileInDirectory[] = [];
+  const maxConcurrent = 50;
```
Why do we have `maxConcurrent` set as 50 here and 25 in the other file?

ipapast replied:
there is a rate limit configured in registry, and we experienced some 502s coming back when we issued too many requests in parallel. We were basically playing with the numbers and running it locally.

So the goal here was to avoid any errors caused by rate limits, just _**slowly**_ work through them.

50 for detecting the files in the directory (we were also considering CPU usage), 25 mostly because of the RATE_LIMIT since we are hitting the endpoint to sendTestPayload() in the for...loop of all the payloads

aron replied:
I would agree that we should keep it consistent, let's use 25. I don't think it's a big deal as the difference will be small, the main thing is that it is batched and doesn't try to process _all_ the files at once.

ipapast replied:
I think we had a conversation about the files and laptops/local configurations limitations, but happy to change it for consistency and re-consider in the future if needed.

JackuB replied:
50 concurrent requests sounds a bit hardcore for Node.js, no?

ipapast replied:
**25, apologies

JackuB reviewed (changes requested):
Looking good, I'd like to clear the questions about swapping of proxy dependencies first.

Also, we do collect some performance metrics from the CLI now, so we should be able to see real-world impact from this change once it's released.

ipapast commented:
> Looking good, I'd like to clear the questions about swapping of proxy dependencies first.
> 
> Also, we do collect some performance metrics from the CLI now, so we should be able to see real-world impact from this change once it's released.

Hi @JackuB, we reverted back to http-proxy-agent and added a comment to explain.
Also went back to 25 concurrent requests in the queue for both detection and running the tests.

Could you check it one final time please? 🙇 
Would you also be able to point me on where to find these performance metrics from the CLI you mentioned?

snyksec commented:
:tada: This PR is included in version 1.438.0 :tada:

The release is available on:
- [npm package (@latest dist-tag)](https://www.npmjs.com/package/snyk/v/1.438.0)
- [GitHub release](https://github.com/snyk/snyk/releases/tag/v1.438.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
